hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/HazelcastSqlValidator.java

```diff
@@ -17,53 +17,284 @@
 import java.util.Set;
 
+import static com.hazelcast.sql.impl.calcite.validate.SqlNodeUtil.isLiteral;
+import static com.hazelcast.sql.impl.calcite.validate.SqlNodeUtil.isParameter;
+import static com.hazelcast.sql.impl.calcite.validate.SqlNodeUtil.numericValue;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.canRepresent;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.isInteger;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.isNumeric;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.narrowestTypeFor;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.typeName;
+import static org.apache.calcite.sql.type.SqlTypeName.CHAR;
+import static org.apache.calcite.sql.type.SqlTypeName.DECIMAL;
+import static org.apache.calcite.sql.type.SqlTypeName.DOUBLE;
+import static org.apache.calcite.sql.type.SqlTypeName.NULL;
+import static org.apache.calcite.sql.type.SqlTypeName.NUMERIC_TYPES;
+import static org.apache.calcite.sql.type.SqlTypeName.VARCHAR;
+import static org.apache.calcite.util.Static.RESOURCE;
+
 /**
  * Hazelcast-specific SQL validator.
  */
 public class HazelcastSqlValidator extends SqlValidatorImpl {
-    public HazelcastSqlValidator(
-        SqlOperatorTable opTab,
-        SqlValidatorCatalogReader catalogReader,
-        RelDataTypeFactory typeFactory,
-        SqlConformance conformance
-    ) {
-        super(opTab, catalogReader, typeFactory, Config.DEFAULT.withSqlConformance(conformance));
+
+    private static final Config CONFIG = Config.DEFAULT.withIdentifierExpansion(true);
+
+    private final Map<SqlNode, RelDataType> knownNodeTypes = new IdentityHashMap<>();
+
+    public HazelcastSqlValidator(SqlOperatorTable opTab, SqlValidatorCatalogReader catalogReader,
+                                 HazelcastTypeFactory typeFactory, SqlConformance conformance) {
+        super(opTab, catalogReader, typeFactory, CONFIG.withSqlConformance(conformance));
+        setTypeCoercion(new HazelcastTypeCoercion(this));
+    }
+
+    /**
+     * Sets {@code type} as the known type for {@code node}.
+     *
+     * @param node the node to set the known type of.
+     * @param type the type to set the know node type to.
+     */
+    public void setKnownNodeType(SqlNode node, RelDataType type) {
+        assert !getUnknownType().equals(type);
+        knownNodeTypes.put(node, type);
+    }
+
+    /**
+     * Obtains a type known by this validator for the given node.
+     *
+     * @param node the node to obtain the type of.
+     * @return the node type known by this validator or {@code null} if the type
+     * of the given node is not known yet.
+     */
+    public RelDataType getKnownNodeType(SqlNode node) {
+        return knownNodeTypes.get(node);
+    }
+
+    @Override
+    public void validateQuery(SqlNode node, SqlValidatorScope scope, RelDataType targetRowType) {
```

Comment:
This method does nothing in the current PR since we do not support LIMIT/OFFSET. 
I would propose to remove it for now and keep only in the `sql` branch because from the 4.1 perspective this is a not-tested dead code.

Reply:
fixed in 255db16e20283e3d3aa8784f19d4d252b5345d80